### PR TITLE
Refactor "new action" logic in ReportActionsView to fix new marker

### PR DESCRIPTION
**Pull request by marcaaron**
### Fixed Issues

$ https://github.com/Expensify/App/issues/11726

### Tests

Same as QA

- [ ] Verify that no errors appear in the JS console

### QA Steps

**New message marker set correctly when report is default while signing in**

1. Open two sessions one as User A and one as User B
1. Make sure User B is navigated to the chat with User A
1. Sign out User B 
1. Create message as User A on chat with User B and wait 10 seconds
1. Sign in User B
1. Verify the chat with User A loads
1. Verify that new marker is set correctly
1. Verify that the chat in the LHN is marked as read 

**New message marker set correctly when navigating to unread chat after signing in**

1. Open two sessions one as User A and one as User B
1. Make sure User B is navigate to a chat other than the chat with User A
1. Sign out User B 
1. Create message as User A on chat with User B and wait 10 seconds
1. Sign in User B
1. Navigate to chat with User A as User B
1. Verify that new marker is set correctly
1. Verify that the chat in the LHN is marked as read 

**New message marker set correctly when navigating to unread chat after opening app**

1. Open two sessions one as User A and one as User B
1. Make sure User B is navigate to a chat other than the chat with User A
1. Close tab (or kill app) with User B (Do not sign out)
1. Create message as User A on chat with User B and wait 10 seconds
1. Open a new tab (or app) as User B
1. Navigate to chat with User A as User B
1. Verify that new marker is set correctly
1. Verify that the chat in the LHN is marked as read 

**New message marker set correctly when navigating to directly to unread chat**

1. Open two sessions one as User A and one as User B
1. Make sure User B is navigate to the chat with User A and save the URL
1. Close User B tab (or kill app)
1. Create message as User A on chat with User B and wait 10 seconds
1. Open tab (or app) as User B using the URL saved from the second step (on native use a deep link)
1. See that new marker is set correctly
1. Verify that the chat in the LHN is marked as read 

- [ ] Verify that no errors appear in the JS console

#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots

QA testing is intense for these flows as there are 4 test cases to look at. I'm only posting videos for 1 scenario, but they should all be tested.

#### Web

https://user-images.githubusercontent.com/32969087/199356480-7810b8b7-1875-420d-9284-2b00340534e9.mp4

#### Mobile Web - Chrome

https://user-images.githubusercontent.com/32969087/199363242-3cb5aaff-867e-4404-bdc4-2ec1df226e2f.mp4

#### Mobile Web - Safari

https://user-images.githubusercontent.com/32969087/199362322-ad1687d8-548d-4d63-b298-e557b14ded8d.mp4

#### Desktop

https://user-images.githubusercontent.com/32969087/199357508-50fa3b32-d6b3-4249-bc7f-99c5c7d06de7.mp4

#### iOS

https://user-images.githubusercontent.com/32969087/199361050-2a2165d1-4a72-4f01-af17-35ded739cf6c.mp4

#### Android

https://user-images.githubusercontent.com/32969087/199363003-5cc88612-e936-4897-a199-87a47afb5344.mp4

### CODE REVIEW

File: src/libs/actions/Report.js [hunk 1]
```diff
@@ -1384,6 +1384,27 @@ function setIsComposerFullSize(reportID, isComposerFullSize) {
...
 
+const defaultNewActionSubscriber = {
+    reportID: '',
+    callback: () => {},
+};
+
+let newActionSubscriber = defaultNewActionSubscriber;
+
+/**
+ * Enables the Report actions file to let the ReportActionsView that a new comment has arrived in realtime for the current report
```
Comment (parasharrajat) [inline src/libs/actions/Report.js:1395]:
```suggestion
 * Enables the Report actions file to let the ReportActionsView know that a new comment has arrived in realtime for the current report
```

File: src/pages/home/ReportScreen.js [hunk 2]
```diff
@@ -217,6 +217,9 @@ class ReportScreen extends React.Component {
         const addWorkspaceRoomOrChatPendingAction = lodashGet(this.props.report, 'pendingFields.addWorkspaceRoom') || lodashGet(this.props.report, 'pendingFields.createChat');
         const addWorkspaceRoomOrChatErrors = lodashGet(this.props.report, 'errorFields.addWorkspaceRoom') || lodashGet(this.props.report, 'errorFields.createChat');
         const screenWrapperStyle = [styles.appContent, styles.flex1, {marginTop: this.state.viewportOffsetTop}];
+
+        // There are reportActions at all to display and we are still in the process of loading the next set of actions.
```
Comment (parasharrajat) [inline src/pages/home/ReportScreen.js:221]:
```suggestion
        // There are no reportActions at all to display and we are still in the process of loading the next set of actions.
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by Beamanator** [comment]
Sorry didn't get to review today, will prioritize this tomorrow

**Comment by marcaaron** [comment]
Updated

**Comment by parasharrajat** [comment]
Testing...

**Review by Beamanator — APPROVED** [review]
Looking good & testing well for me, I like the loads of comments since we're doing some funky stuff, but overall it's making sense 👍 Won't merge till @parasharrajat does his review

**Comment by parasharrajat** [comment]
I am facing some issues with Mac so I won't be able to test all platforms until that issue is fixed. 

Hoping to get it fixed asap and get back on it.
- Thanks

**Comment by marcaaron** [comment]
@Beamanator Can you jump in here and help test on all platforms? We need at least one person to post the checklist.

**Comment by Beamanator** [comment]
Fak I'm no longer at my laptop for about 5 hours, sorry - can test Sunday if that's ok for you

**Comment by Beamanator** [comment]
<details>
<summary><h4>IOS App w/ Chrome</h4></summary>

When logging out / logging in:

https://user-images.githubusercontent.com/3885503/200322316-617d5b4b-d80f-4440-b9f0-af56fe3c7d71.mov

When closing & opening app (notice the new message shows in iOS LHN, when opening the report it is shown a bit late w/ an animation - is that on purpose?):

https://user-images.githubusercontent.com/3885503/200322543-b0948d79-7141-4cee-a9f8-b54a58acfc91.mov

</details>

<details>
<summary><h4>iOS Safari w/ Android</h4></summary>

When logging out / logging in:

https://user-images.githubusercontent.com/3885503/200324665-168aa41a-01da-4588-9f4a-a827e4412cd5.mov

When closing & opening app (looks like iOS Safari mWeb didn't get the "unread" marker 🤔 :

https://user-images.githubusercontent.com/3885503/200325019-8bd1528b-731a-4535-91b3-07714951970e.mov

but it worked this way??

https://user-images.githubusercontent.com/3885503/200326555-abf85375-b37b-4408-bbaa-a5ec1c327e47.mov


</details>

<details>
<summary><h4>Android Chrome w/ Desktop</h4></summary>

When logging out / logging in:

https://user-images.githubusercontent.com/3885503/200326163-dd7a6180-29ae-421e-b372-46aa572d7f05.mov

When closing & opening app:

https://user-images.githubusercontent.com/3885503/200326363-2f0fe37a-3df2-41c2-95d9-1e8d990b2444.mov

</details>

**Comment by marcaaron** [comment]
Thanks for testing @Beamanator !

>notice the new message shows in iOS LHN, when opening the report it is shown a bit late w/ an animation - is that on purpose?

Expected, yes! It's just a consequence of how we load reports first then load comments second when tapping into the chat. There are some discussions floating around about how to improve this IIRC.

>When closing & opening app (looks like iOS Safari mWeb didn't get the "unread" marker

Hmm hard to say, but I think I'd expect that to work. I probably would have rejected this PR for not passing the test steps. But we can see if it happens for QA and if a new issue needs to be created we'll look into it.

**Comment by marcaaron** [comment]
Heads up @Beamanator @parasharrajat elements of this PR were reverted because they caused a pretty nasty regression related to offline behavior. I think we can all learn from this PR by making sure to test offline behaviors and include them in the test steps for our future PRs. Thanks.

**Comment by parasharrajat** [comment]
Sadly, I didn't get the chance to test it. But I agree offline is a must for this PR.

**Comment by marcaaron** [comment]
offline testing is a must for ~~this PR~~ all PRs ? 😄

**Comment by parasharrajat** [comment]
Correct :sweat_smile: . Saw your update.

**Pull request merged**